### PR TITLE
User accounts

### PR DESCRIPTION
**Pull request by monkeydom**
I think this would qualify for prelminary user accounts and good to merge in before the addition of permissions to keep the changes small

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -61,6 +66,11 @@ defmodule Radiator.MixProject do
       {:poison, "~> 3.0"},
       {:timex, "~> 3.5"},
       {:cors_plug, "~> 2.0"},
+      # authentication
+      {:guardian, "~> 1.2"},
+      {:argon2_elixir, "~> 2.0"},
```
Comment (jeyemwey) [inline mix.exs]: `argon2_elixir` [requires](https://github.com/riverrun/comeonin/wiki/Requirements) gcc, make and glibc to build.

We could either:
1. ...make the C compiler a requirement for radiator-spark and define it as such in the Readme. Comeonin has this paragraph in its Readme:
  > Make sure you have a C compiler installed. See the [Comeonin wiki](https://github.com/riverrun/comeonin/wiki/Requirements) for details.
2. ...choose an other password hashing algorithm which is implemented in Elixir natively and does not rely on gcc.

While installing C on *nix systems is not that complicated, it is a larger act on Windows systems.

File: lib/radiator_web/controllers/login_controller.ex [hunk 2]
```diff
@@ -0,0 +1,137 @@
+defmodule RadiatorWeb.LoginController do
```
Comment (eteubert) [inline lib/radiator_web/controllers/login_controller.ex]: It feels to me like this controller has to many responsibilities. Also, if I can, I try to stay to the "standard" controller actions.

An abstraction I found useful is to model login around "sessions" as this is an object that corresponds well with the standard controller actions: new (form), create (login/auth action), delete (logout).

Signup should be in a separate RegistrationController. Even the verification flow might benefit from being extracted into its own controller.

File: mix.exs (hunk 1 above)
Comment (monkeydom) [reply mix.exs]: I would go for 1) - as we probably won't be able to just rely on native elixir libraries anyways. E.g. as soon as we do any kind of audio processing I think we need to rely on c code. @eteubert what do you think?

Comment (eteubert) [reply mix.exs]: A valid concern but I also think that for our project scope we will likely need more libraries that depend on a C compiler later.

File: lib/radiator/auth/directory.ex [hunk 3]
```diff
@@ -0,0 +1,49 @@
+defmodule Radiator.Auth.Directory do
```
[the remark this answers is not in the document]
Comment (monkeydom) [reply lib/radiator/auth/directory.ex]: Yeah - already ran into that. Did not have a better name in my head though, any suggestions? I could just make it the Main Radiator.Auth module, but don't know if that is good style either.

File: lib/radiator_web/templates/login/login_form.html.eex [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<div class="card">
+<%= form_for @conn, @action, [multipart: true], fn f -> %>
```
[the remark this answers is not in the document]
Comment (monkeydom) [reply lib/radiator_web/templates/login/login_form.html.eex]: Ok, will do. Just copy and pasted and wanted to keep sure this is a post request to not generate a query string, but the default already does this.

File: lib/radiator/auth/directory.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (eteubert) [reply lib/radiator/auth/directory.ex]: On second look, I also question if "Auth" is a good choice here as top level namespace. For example `&Radiator.Auth.Directory.get_user_by_email/1` is not really part of authentication, it's a User/Account helper function, right?

Would it possibly make sense to turn this around and introduce a top level generic `Radiator.Account` context that for example provides `&Radiator.Account.get_user_by_email/1`, the User under `Radiator.Account.User` and Auth concerns nested under `Radiator.Account.Auth`?

File: lib/radiator/auth/user.ex [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      changeset
+    end
+  end
+
+  def check_password(%User{} = user, password) do
+    Argon2.check_pass(user, password)
+  end
+
+  @request_email_salt "34ercft gpaojrt[we"
+  @email_verification_salt "q3480haspodfinp0284;a"
```
[the remark this answers is not in the document]
Comment (monkeydom) [reply lib/radiator/auth/user.ex]: Eventually maybe. For now I don't think so.

I am using the Phoenix token library with the endpoint key-base for them. 

The first one is just to make sure only logged in users can request this, so no easy spam is possible (although we also need to provide a recovery for forgotten emails eventually), for the verification email I also encode the the username/email combo in it.

E.g. I think the salts here just serve the purpose to ensure that the same payload for differrent mechanism doesn't result in identical tokens, therefore just having them be different for each use case should be enough.

File: lib/radiator_web/controllers/login_controller.ex (hunk 2 above)
Comment (monkeydom) [reply lib/radiator_web/controllers/login_controller.ex]: Okay, in that model though if I would want a combined login/signup page, who would be the server of that?

Not sure if I like the idea of the session being a resource style thing, as it is fleeting and not permanent, which would be the separation in my head. Also isn't that misleading? We do have a session for each contact, regardless if they are logged in or not, don't we?

Comment (eteubert) [reply lib/radiator_web/controllers/login_controller.ex]: Hm true, technically the session is there even for logged out users, the login action only _alters_ the session.. Then ... `AuthenticationController` maybe? But then you could also leave it as `LoginController`.

Regarding combined pages: Not sure, possibly another controller just for rendering the combined form/page? The actual login/register actions I would still keep in separate controllers.

File: lib/radiator/auth/directory.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (monkeydom) [reply lib/radiator/auth/directory.ex]: Well, I do think the user table is part of the Authentication and keep it part of it. E.g. have a separation between the data model Radiator is using in general for users, and the data needed to have and manage Authentication. So in my head there is a user/account idea in Radiator direct too, but that would store settings, profile data like avatars etc. 

Ideally this authenticated user can be used for many different separate functionalities.

[the remark this answers is not in the document]
Comment (eteubert) [reply lib/radiator/auth/directory.ex]: Then I think your suggestion from above is good, making `Radiator.Auth` the context.

[the remark this answers is not in the document]
Comment (monkeydom) [reply lib/radiator/auth/directory.ex]: Renamed it to Register for now to at least avoid the name clash.

[the remark this answers is not in the document]
Comment (eteubert) [reply lib/radiator/auth/directory.ex]: Not in love with it but it solves the main issue 👍

**Pull request merged**